create/modify all history events (#6700)
rotki/rotki: docs/api.rst

                  "sequence_index": 162,
                  "timestamp": 1569924574,
                  "location": "ethereum",
                  "event_type": "informational",
                  "asset": "eip155:1/erc20:0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359",
                  "balance": {"amount": "1.542", "usd_value": "1.675"},
                  "location_label": "0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12",
                  "notes": "Approve 1 SAI of 0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12 for spending by 0xdf869FAD6dB91f437B59F1EdEFab319493D4C4cE",
                  "event_subtype": "approve",
                  "counterparty": "0xdf869FAD6dB91f437B59F1EdEFab319493D4C4cE",
                  "extra_data": {}
               }

            :reqjson string tx_hash: This is the transaction hash of the evm event
            :reqjson int sequence_index: This is an index that tries to provide the order of history entries for a single event_identifier.
            :reqjson string location: The location of the entry. Such as "ethereum", "optimism", etc.
            :reqjson string asset: The asset identifier for this entry

>             :reqjson string event_type: The main event type of the entry. Possible event types can be seen in HistoryEventType enum.

Author: that would probably mean I create a separate section just for these two arguments? I'm not sure what you have in mind

>             :reqjson string event_type: The main event type of the entry. Possible event types can be seen in HistoryEventType enum.

Reviewer: No I mean add a link to the definition ofthe enum.

               "balance": {"amount": "1.542", "usd_value": "1.675"},
               "sequence_index": 162,
               "event_type": "informational",
               "event_subtype": "approve",
               "asset": "eip155:1/erc20:0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359",
               "location_label": "0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12",
               "notes": "Approve 1 SAI of 0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12 for spending by 0xdf869FAD6dB91f437B59F1EdEFab319493D4C4cE"
            }

         :reqjson int sequence_index: This is an index that tries to provide the order of history entries for a single event_identifier.
         :reqjson string location: The location of the entry. Such as "ethereum", "optimism", etc.
         :reqjson string asset: The asset identifier for this entry
         :reqjson string event_identifier: The event identifier to be used for the event.
         :reqjson string event_type: The main event type of the entry. Possible event types can be seen in the `HistoryEventType enum <https://github.com/rotki/rotki/blob/59aa288dacd1776e62682e711a916f32a14c04c2/rotkehlchen/accounting/structures/types.py#L54>`_.

>          :reqjson string[optional] event_subtype: An optional subtype for the entry. Possible event types can be seen in the `HistoryEventSubType enum <https://github.com/rotki/rotki/blob/59aa288dacd1776e62682e711a916f32a14c04c2/rotkehlchen/accounting/structures/types.py#L72>`_.

Reviewer: Same here. Why are you duplicating those btw? Shouldn't they be in the base description of the hisory event only?

```suggestion
         :reqjson event_subtype: The subtype for the entry. Possible event types can be seen in the `HistoryEventSubType enum <https://github.com/rotki/rotki/blob/59aa288dacd1776e62682e711a916f32a14c04c2/rotkehlchen/accounting/structures/types.py#L72>`_.
```

Author: they only appear in two of the subclasses. `entry_type`, `timestamp` & `balance` are the only fields that apply to all subclasses

      **Example Request**:

      .. http:put:: /api/(version)/history/events

         .. http:example:: curl wget httpie python-requests

            PUT /api/1/history/events HTTP/1.1
            Host: localhost:5042
            Content-Type: application/json;charset=UTF-8

            {
               "entry_type": "eth withdrawal event",
               "timestamp": 1569924574,
               "balance": {"amount": "1.542", "usd_value": "1.675"},
               "is_exit": true,
               "validator_index": 1,
               "withdrawal_address": "0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12"
            }

         :reqjson int validator_index: This is the index of the validator.
         :reqjson string withdrawal_address: an evm address field to specify the withdrawer in an "eth withdrawal event".
         :reqjson bool is_exit: true if the "eth withdrawal event" is an exit event.

   :reqjson string entry_type: The type of the event that will be processed. Different validation is used based on the value for this field. Possible values are: ``"history event"``, ``"evm event"``, ``"eth withdrawal event"``, ``"eth block event"``, ``"eth deposit event"``.

>    :reqjson int timestamp: The timestamp of the entry

Reviewer: Please specify here that it's in milliseconds now

          "location_label": "0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12",
          "notes": "Approve 1 SAI of 0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12 for spending by 0xdf869FAD6dB91f437B59F1EdEFab319493D4C4cE",
          "event_subtype": "approve",
          "counterparty": "0xdf869FAD6dB91f437B59F1EdEFab319493D4C4cE",
          "extra_data": {}
      }

   .. http:example:: curl wget httpie python-requests

      PUT /api/1/history/events HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

      {
          "entry_type": "history event",
          "event_identifier": "RE_xxxxxxxxxx",
          "location": "ethereum",
          "timestamp": 1569924574,
          "balance": {"amount": "1.542", "usd_value": "1.675"},
          "sequence_index": 162,
          "event_type": "informational",
          "event_subtype": "approve",
          "asset": "eip155:1/erc20:0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359",
          "location_label": "0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12",
          "notes": "Approve 1 SAI of 0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12 for spending by 0xdf869FAD6dB91f437B59F1EdEFab319493D4C4cE"

>       }

Reviewer 2: I'm not sure but is it possible to create sub sections inside this section for the different event types and then have each one with their special fields only and then later the fields that are common for all with a reference? I believe it would improve this section a lot

          "is_mev_reward": true
      }

   .. http:example:: curl wget httpie python-requests

      PUT /api/1/history/events HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

      {
          "entry_type": "eth deposit event",
          "timestamp": 1569924574,
          "balance": {"amount": "1.542", "usd_value": "1.675"},
          "tx_hash": "0x64f1982504ab714037467fdd45d3ecf5a6356361403fc97dd325101d8c038c4e",
          "event_identifier": "RE_xxxxxxxxxx",
          "validator_index": 1,
          "sequence_index": 162,
          "depositor": "0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12",
          "extra_data": {}
      }

   .. http:example:: curl wget httpie python-requests

      PUT /api/1/history/events HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

      {
          "entry_type": "eth withdrawal event",
          "timestamp": 1569924574,
          "balance": {"amount": "1.542", "usd_value": "1.675"},
          "is_exit": true,
          "validator_index": 1,
          "withdrawal_address": "0x2B888954421b424C5D3D9Ce9bB67c9bD47537d12"
      }

>    :reqjson string entry_type: This is the type of the event.

Reviewer 2: ```suggestion
   :reqjson string entry_type: The type of the event that will be processed. Different validation is used based on the value for this field
```

also list the possible values

   :reqjson string tx_hash: This is the transaction hash of the evm event
   :reqjson int sequence_index: This is an index that tries to provide the order of history entries for a single event_identifier.

>    :reqjson int timestamp: The timestamp of the entry

Reviewer 2: Don't forget to also split the arguments depending on the type. As it is now is unredeable

   :reqjson string notes: This is a description of the event entry in plain text explaining what is being done. This is supposed to be shown to the user.
   :reqjson string event_subtype: Optional. An optional subtype for the entry. Possible event types can be seen in HistoryEventSubType enum.
   :reqjson string counterparty: Optional. An identifier for a potential counterparty of the event entry. For a send it's the target. For a receive it's the sender. For bridged transfer it's the bridge's network identifier. For a protocol interaction it's the protocol.

>    :reqjson int block_number: This is the index of the block the event is contained in.

Reviewer 2: ```suggestion
   :reqjson int block_number: This is the number of the block where the event took place.
```

>    :reqjson int validator_index: This is the index of the validator.

Reviewer 2: can you check the errors? I bleieve the text need updates. For example if there is a constraint that fails like the validator index doesn't exists, I believe there is one like that

Author: The errors seem correct

>    :reqjson string event_identifier: The event identifier to be used for the event.

Reviewer 2: This one has to be unique too

   :reqjson string fee_recipient: an evm address field to specify the recipient in an "eth block event".
   :reqjson string depositor: an evm address field to specify the depositor in an "eth deposit event".
   :reqjson string withdrawal_address: an evm address field to specify the withdrawer in an "eth withdrawal event".

>    :reqjson bool is_exit: boolean field to specify if the "eth withdrawal event" is an exit.

Reviewer 2: ```suggestion
   :reqjson bool is_exit: true if the "eth withdrawal event" is an exit event.
```

>    :reqjson bool is_exit: boolean field to specify if the "eth withdrawal event" is an exit.

Reviewer 2: ```suggestion
   :reqjson bool is_exit: true if the "eth withdrawal event" is an exit event.
```

>    :reqjson bool is_mev_reward: boolean field to specify if the "eth block event" is an mev reward event.

Reviewer 2: ```suggestion
   :reqjson bool is_mev_reward: true if the "eth block event" is an mev reward event.
```
